Marlin/Marlin_main.cpp

```diff
@@ -9946,42 +9946,23 @@ void quickstop_stepper() {
 #if HAS_M206_COMMAND
 
   /**
-   * M428: Set home_offset based on the distance between the
-   *       current_position and the nearest "reference point."
-   *       If an axis is past center its endstop position
-   *       is the reference-point. Otherwise it uses 0. This allows
-   *       the Z offset to be set near the bed when using a max endstop.
-   *
-   *       M428 can't be used more than 2cm away from 0 or an endstop.
+   * M428: Set a persistent offset to the native home position and coordinate
+   *       space by assigning the current position as the native home position.
    *
    *       Use M206 to set these values directly.
    */
   inline void gcode_M428() {
-    bool err = false;
-    LOOP_XYZ(i) {
-      if (axis_homed[i]) {
-        const float base = (current_position[i] > (soft_endstop_min[i] + soft_endstop_max[i]) * 0.5) ? base_home_pos((AxisEnum)i) : 0,
-                    diff = base - current_position[i];
-        if (WITHIN(diff, -20, 20)) {
-          set_home_offset((AxisEnum)i, diff);
-        }
-        else {
-          SERIAL_ERROR_START();
-          SERIAL_ERRORLNPGM(MSG_ERR_M428_TOO_FAR);
-          LCD_ALERTMESSAGEPGM("Err: Too far!");
-          BUZZ(200, 40);
-          err = true;
-          break;
-        }
-      }
-    }
+    stepper.synchronize();
+    if (axis_unhomed_error())
+      return;
 
-    if (!err) {
-      report_current_position();
-      LCD_MESSAGEPGM(MSG_HOME_OFFSETS_APPLIED);
-      BUZZ(100, 659);
-      BUZZ(100, 698);
-    }
+    LOOP_XYZ(i)
+      set_home_offset((AxisEnum)i, -1.0f * current_position[i]);
```

Comment:
The problem with this is that it doesn't account for min versus max endstops. The "home position" for a max endstop might be something like 300. Simply setting the home offset based on the current position assumes that the home position set in the configuration is always going to be zero. The logic needs to account for the difference between the current position and the endstop — so the current implementation simply works from the distance between the current position and the nearest end of the axis.